crates/api/src/local_user/toggle_totp.rs

```diff
@@ -0,0 +1,61 @@
+use crate::check_totp_2fa_valid;
+use actix_web::{
+  web::{Data, Json},
+  HttpResponse,
+};
+use lemmy_api_common::{context::LemmyContext, person::ToggleTotp};
+use lemmy_db_schema::{
+  source::local_user::{LocalUser, LocalUserUpdateForm},
+  traits::Crud,
+};
+use lemmy_db_views::structs::{LocalUserView, SiteView};
+use lemmy_utils::error::{LemmyError, LemmyErrorType};
+
+/// Enable or disable two-factor-authentication. The current setting is determined from
+/// [LocalUser.totp_2fa_enabled].
+///
+/// To enable, you need to first call [generate_totp_secret] and then pass a valid token to this
+/// function.
+///
+/// Disabling is only possible if 2FA was previously enabled. Again it is necessary to pass a valid
+/// token.
+#[tracing::instrument(skip(context))]
+pub async fn toggle_totp(
+  data: Json<ToggleTotp>,
+  local_user_view: LocalUserView,
+  context: Data<LemmyContext>,
+) -> Result<HttpResponse, LemmyError> {
+  let site_view = SiteView::read_local(&mut context.pool()).await?;
+
+  // require valid 2fa token to enable or disable 2fa
+  if local_user_view.local_user.totp_2fa_secret.is_none() {
+    return Err(LemmyErrorType::MissingTotpToken.into());
+  }
+  check_totp_2fa_valid(
```

Comment:
This will early return / fail if the `totp_token` is an empty or invalid string (assuming that's how to disable it), and you won't be able to disable it.

So it really should be `let new_totp_state = if let Ok() ... set enabled to true } else { set enabled to false};`

Reply:
Also, the `check_totp_2fa_valid` function just returns `Ok(())` if its disabled, meaning it won't actually check anything. 

So there probably should be a `check_totp_2fa_toggle` and `check_totp_2fa_login`, where the latter one cares if you have it enabled, but the first one doesn't, and checks it every time.

Reply:
Fixed by moving the 2fa enabled check out of `check_totp_2fa_valid` into `login.rs`. Btw the token definitely needs to be passed in order to disable 2fa, otherwise this could be used by an attacker to bypass 2fa.